sdks/python/apache_beam/io/gcp/bigquery_tools.py

```diff
@@ -1491,7 +1493,29 @@ class RetryStrategy(object):
   RETRY_NEVER = 'RETRY_NEVER'
   RETRY_ON_TRANSIENT_ERROR = 'RETRY_ON_TRANSIENT_ERROR'
 
-  _NON_TRANSIENT_ERRORS = {'invalid', 'invalidQuery', 'notImplemented'}
+  # Values below may be found in reasons provided either in an
+  # error returned by a client method or by an http response as
+  # defined in google.api_core.exceptions
+  _NON_TRANSIENT_ERRORS = {
+      'invalid',
+      'invalidQuery',
+      'notImplemented',
+      'Moved Permanently',
+      'Not Modified',
+      'Temporary Redirect',
+      'Resume Incomplete',
+      'Bad Request',
+      'Unauthorized',
+      'Forbidden',
+      'Not Found',
+      'Method Not Allowed',
+      'Conflict',
+      'Length Required',
+      'Precondition Failed',
+      'Not Implemented',
+      'Bad Gateway',
+      'Gateway Timeout',
+  }
```

Comment:
> Thanks for adding this comprehensive list! I'm a little unfamiliar with a few of these so will need another pair of eyes to check.
> 
> From a cursory search though, it appears "bad gateway" may be a transient error? Same with "gateway timeout". I would defer to https://cloud.google.com/bigquery/docs/error-messages for a good list to take from.

I see what you mean about some of the one I added possibly being transient, we certainly could take some of the 5XX errors out of that list to broaden it a bit.

I went narrow, initially, based on the narrow set of api core exceptions considered transient in api core retry [here](https://github.com/googleapis/python-api-core/blob/main/google/api_core/retry.py#L125-L128), where it's only 429 Too Many Requests, 500 Internal Server Error, and 503 Service Unavailable that are considered transient. Other than those 3, all remaining 4XX, 5XX, and 3XX errors are not considered transient.

Let me know if you think we should remove some 5XX exceptions from the list or if we should stick with just retrying 429, 500, and 503.

Reply:
> Also error reasons from BQ seem to appear in camel case format (e.g. `notFound` instead of `Not Found`). Is it different for GRPC errors?

Posted this in a different comment but will include here to make it easy to follow in this thread - 

There are 2 types of error reasons in that list:
1. Reasons returned in the error list response of the `self.gcp_bq_client.insert_rows_json` call. Those are in camel case and are the ones found here in this doc - https://cloud.google.com/bigquery/docs/error-messages.
2. Reasons returned by the `GoogleAPICalError` exceptions. Those follow the HTTP standard reason format. A 404, for example will return a reason of `Not Found`, which is verified with [this integration test](https://github.com/apache/beam/pull/28091/files#diff-b6e14063dc720a5be1a68c935803f399ab5301bc3748c7f46f156b0c80f5d691R484).

Reply:
Gotcha, I'll review that and trim that list down a bit, seems like retrying more of the 5XX errors would make sense.

Reply:
Added [this commit](https://github.com/apache/beam/pull/28091/commits/ec629474e8539b9ca577eefccf8623a503df343e) which removed a few exceptions based on our conversation and a review of the [bq error table](https://cloud.google.com/bigquery/docs/error-messages#errortable):

- Errors that don't seem to be relevant:
  - 3XX Errors - Moved Permanently, Not Modified, Temporary Redirect, Resume Incomplete
  - Some 4XX Errors - Method Not Allowed, Length Required, Precondition Failed
- Errors that should be retried:
  - 409 Conflict
  - 502 Bad Gateway
  - 504 Gateway Timeout